src/sinks/util/tcp.rs

```diff
@@ -299,7 +304,7 @@ where
                 if error.kind() == ErrorKind::Other && error.to_string() == "ShutdownCheck::Close" {
                     emit!(TcpSocketConnectionShutdown {});
```

Comment:
Should `TcpSocketConnectionShutdown` also emit an `EventsDropped`? 

I'm guessing it's a fairly catastrophic error, but we do keep looping around the events trying to push the events to the closed server. I presume all those events will be lost. 

Should it actually break out here and shutdown the sink?

Reply:
Goood question.
I almost added an EventsDropped to `TcpSocketConnectionShutdown` , but I hesitated because that event currently only does this:

```

Reply:
Reading into some further comments in this code:

```
/// Call `shutdown_check` on each `poll_flush`, so we can stop sending data if other side disconnected.
```

```
                  // Maybe this is only a sign to close the channel,                                                                                                                                                                                                                                                                                  
                  // in which case we should try to flush our buffers                                                                                                                                                                                                                                                                                 
                  // before disconnecting.
```

It looks to me like this case of the buffer being empty was specifically coded to avoid emitting `EventsDropped` and `Error` . Kind of looks like the idea is that it is potentially a hint that errors might come, but that we'll keep trying and if those next events result in a true tcp error, then we fire `TcpSocketSendError` for each one.
:shrug: 
But, I could definitely be wrong.

Reply:
@StephenWakely  , I reviewed this logic again I'm still on the fence :smile: 

The follow code is where the case `ShutdownCheck::Close` is determined, see the comment there.

https://github.com/vectordotdev/vector/blob/f32eb264184bcee4f635984f7e7fef22dd741380/src/sinks/util/tcp.rs#L236-L258

It is a state that we return when the socket read operation did not return an explicit error, but it returned `Ok` and zero bytes were read.

I guess the point I'm trying to make is- I'm convinced that the existing logic to not fire `EventsDropped` nor `Error` internal events when `ShutdownCheck::Close` condition occurs, is correct - _if the original author's code/comments are correct_ (if the `Ok(0)` condition is an indication that the connection might still be open, so we keep trying to flush the buffer until we get an explicit error).

My hesitation is that this is currently a debug! message, and if we decide that `TcpSocketConnectionShutdown` should emit a `EventsDropped` , then that necessitates that it should also be upgraded to emitting an `Error` event as well, and that doing so  might create excess errors?

To this question:

> Should it actually break out here and shutdown the sink?

Based on my reading of this, it looks like we try to re-connect as we process input.

Reply:
Addressed in https://github.com/vectordotdev/vector/pull/14425/commits/63731b8c947ad012f0df79f9489ef0c146b6b271

, based on our broader discussion.